bin/BuildPackages.sh

```diff
@@ -79,115 +135,197 @@ build_carat() {
 # It is not possible to move around compiled binaries because these have the
 # path to some data files burned in.
 zcat carat-2.1b1.tgz | tar pxf -
-ln -s carat-2.1b1/bin bin
+# If the symbolic link does not exist then create it.
+# Does not check if bin exists already with some content or as a file, etc.
+# Thus this code is unstable, but should be corrected by package author
+# and not by this script.
+if [[ ! -L ./bin ]]
+then
+  ln -s ./carat-2.1b1/bin ./bin
+fi
 cd carat-2.1b1
-make TOPDIR=`pwd`
+make TOPDIR="$(pwd)"
 chmod -R a+rX .
 cd bin
-aa=`./config.guess`
-for x in "`ls -d1 $GAPDIR/bin/${aa}*`"; do
- ln -s "$aa" "`basename $x`"
-done
+# This sets GAParch as it should be.
+# Alternatively, one could check $GAPDIR/bin for all directories instead.
+source "$GAPDIR/sysinfo.gap"
+# If the symbolic link does not exist then create it.
+if [[ ! -L ./"$GAParch" ]]
+then
+  shopt -s nullglob
+  # We assume that there is only one appropriate directory....
+  for archdir in *
+  do
+    if [[ -d "$archdir" ]] && [[ ! -L "$archdir" ]]
+    then
+      ln -s ./"$archdir" ./"$GAParch"
+    fi
+  done
+  shopt -u nullglob
+fi
 )
 }
 
 build_cohomolo() {
 (
-./configure $GAPDIR
+./configure "$GAPDIR"
 cd standalone/progs.d
 cp makefile.orig makefile
 cd ../..
-$MAKE
+"$MAKE"
 )
 }
 
 build_fail() {
-  echo "= Failed to build $dir"
+  echo ""
+  warning "Failed to build $PKG"
+  echo "$PKG" >> "$LOGDIR/$FAILPKGFILE.log"
 }
 
 run_configure_and_make() {
   # We want to know if this is an autoconf configure script
   # or not, without actually executing it!
-  if [ -f autogen.sh ] && ! [ -f configure ] ; then
+  if [[ -f autogen.sh ]] && ! [[ -f configure ]]
+  then
     ./autogen.sh
-  fi;
-  if [ -f configure ]; then
-    if grep Autoconf ./configure > /dev/null; then
-      ./configure $CONFIGFLAGS --with-gaproot=$GAPDIR
+  fi
+  if [[ -f "configure" ]]
+  then
+    if grep Autoconf ./configure > /dev/null
+    then
+      ./configure "$CONFIGFLAGS" --with-gaproot="$GAPDIR"
     else
-      ./configure $GAPDIR
-    fi;
-    $MAKE
+      ./configure "$GAPDIR"
+    fi
+    "$MAKE"
   else
-    echo "No building required for $dir"
-  fi;
+    notice "No building required for $PKG"
+  fi
 }
 
-for dir in `ls -d */`
+build_one_package() {
+  # requires one argument which is the package directory
+  PKG="$1"
+  echo ""
+  date
+  echo ""
+  notice "==== Checking $PKG"
+  (  # start subshell
+  set -e
+  cd "$CURDIR/$PKG"
+  case "$PKG" in
+    anupq*)
+      ./configure "CFLAGS=-m32 LDFLAGS=-m32 --with-gaproot=$GAPDIR" && \
```

Comment:
I see you retained the `&&` here after all. So do they serve a purpose? In that case, perhaps add a comment explaining it? Otherwise, perhaps remove them after all?